File: python/2975. Maximum Square Area by Removing Fences From a Field.py
class Solution {
  maximizeSquareArea(m, n, hFences, vFences) {
    // Add boundary fences
    const sortedHFences = [...hFences, 1, m].sort((a, b) => a - b);
    const sortedVFences = [...vFences, 1, n].sort((a, b) => a - b);
    
    // Calculate horizontal gaps
    const hGaps = new Set();
    for (let i = 0; i < sortedHFences.length; i++) {
      for (let j = 0; j < i; j++) {
        hGaps.add(Math.abs(sortedHFences[i] - sortedHFences[j]));
      }
    }
    
    // Calculate vertical gaps
    const vGaps = new Set();
    for (let i = 0; i < sortedVFences.length; i++) {
      for (let j = 0; j < i; j++) {
        vGaps.add(Math.abs(sortedVFences[i] - sortedVFences[j]));
      }
    }
    
    // Find maximum common gap
    let maxGap = -1;
    // Convert to array and sort in descending order
    const sortedHGaps = Array.from(hGaps).sort((a, b) => b - a);
    
    for (const hGap of sortedHGaps) {
      if (vGaps.has(hGap)) {
        maxGap = hGap;
        break;
      }
    }
    
    // Return result
    if (maxGap === -1) return -1;
    return (maxGap * maxGap) % (10**9 + 7);
  }
}

// Test with the example
const solution = new Solution();
const m = 4;
const n = 3;
const hFences = [2, 3];
const vFences = [2];
console.log(solution.maximizeSquareArea(m, n, hFences, vFences)); // Output: 4

File: 2975. Maximum Square Area by Removing Fences From a Field.js

// var maximizeSquareArea = function(m, n, hFences, vFences) {
//     const MOD = 1e9 + 7;
    
//     // Add boundary fences
//     hFences.push(1, m);
//     vFences.push(1, n);
    
//     // Remove duplicates and sort
//     hFences = [...new Set(hFences)].sort((a, b) => a - b);
//     vFences = [...new Set(vFences)].sort((a, b) => a - b);
    
//     // Get all possible horizontal distances
//     const hDistances = new Set();
//     for (let i = 0; i < hFences.length; i++) {
//         for (let j = i + 1; j < hFences.length; j++) {
//             hDistances.add(hFences[j] - hFences[i]);
//         }
//     }
    
//     // Get all possible vertical distances
//     const vDistances = new Set();
//     for (let i = 0; i < vFences.length; i++) {
//         for (let j = i + 1; j < vFences.length; j++) {
//             vDistances.add(vFences[j] - vFences[i]);
//         }
//     }
    
//     // Find maximum common distance
//     let maxCommonDist = 0;
    
//     // Convert hDistances to array for iteration
//     const hDistArray = Array.from(hDistances);
//     for (const dist of hDistArray) {
//         if (vDistances.has(dist) && dist > maxCommonDist) {
//             maxCommonDist = dist;
//         }
//     }
    
//     // Return area modulo MOD, or -1 if no square found
//     if (maxCommonDist === 0) return -1;
//     return Number((BigInt(maxCommonDist) * BigInt(maxCommonDist)) % BigInt(MOD));
// };
var maximizeSquareArea = function(m, n, hFences, vFences) {
    let ret = -1;
    hFences.push(1);
    hFences.push(m);
    vFences.push(1);
    vFences.push(n);

    hFences.sort((a,b) => a - b);
    vFences.sort((a,b) => a - b);

    const s = new Set();
    const bigFences = hFences.length < vFences.length ? vFences : hFences;
    const smallFences = hFences.length < vFences.length ? hFences : vFences;

    for(let i = smallFences.length - 1; i > 0; i--){
        for(let j = i - 1; j >= 0; j--){
            s.add(smallFences[i] - smallFences[j]);
        }
    };

    let maxCommon = -1;
    for(let i = bigFences.length - 1; i > 0; i--){
        for(let j = 0; j < i; j++){
            const d = bigFences[i] - bigFences[j];
            if(d <= maxCommon) break;
            if(s.has(d)){
                maxCommon = Math.max(maxCommon, d);
            }
        }
    };

    if (maxCommon === -1) return -1;

  const MOD = 1000000007n;
  const D = BigInt(maxCommon);
  return Number((D * D) % MOD);

};
// Test with the example
const m = 4;
const n = 3;
const hFences = [2, 3];
const vFences = [2];
console.log(maximizeSquareArea(m, n, hFences, vFences)); // Output: 4